### ISSUE TITLE
Isolate.resolvePackageUri throws uncaught RangeError

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#25955]
In an isolate with a package config that includes a mapping for `path`, The code `Isolate.resolvePackageUri("package:path")` crashes with the following error:

```
RangeError: Value not in range: 5
#0      _StringBase.substring (dart:core-patch/string_patch.dart:365)
#1      _resolvePackageUri (dart:_builtin:298)
#2      _resolvePackageUriFuture.<_resolvePackageUriFuture_async_body> (dart:_builtin:757)
#3      _RootZone.runUnary (dart:async/zone.dart:1137)
#4      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:551)
#5      _Future._propagateToListeners (dart:async/future_impl.dart:637)
#6      _Future._completeWithValue (dart:async/future_impl.dart:424)
#7      _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:479)
#8      _microtaskLoop (dart:async/schedule_microtask.dart:41)
#9      _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
#10     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#11     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:149)
```

I'd expect this to return the URI associated with the `path` package in the config map.

**Comment by lrhn** [comment]
It shouldn't fail that way.
The URI isn't a valid "package:" URI because it doesn't have a slash after the package name, but the VM `builtin.dart` code seems to assume that the slash is there.

What happens if you have an `import "package:path";` in your code?

**Comment by nex3** [comment]
> The URI isn't a valid "package:" URI because it doesn't have a slash after the package name, but the VM `builtin.dart` code seems to assume that the slash is there.

Why isn't it valid? I don't think we have a concrete specification for valid `package:` URIs, and it's pretty clear from the user's perspective that it refers to the directory that contains `path`'s contents. That's how it works if you're using a package root.

> What happens if you have an `import "package:path";` in your code?

It doesn't work. That URI refers to a directory, not a file.

**Comment by lrhn** [comment]
URIs have a nice and (somewhat) consistent specification. Package-URIs are ... not matching that specification very well.

I have tried to specify how a package URI is interpreted for the package-spec DEP and in the resource package (although I'm not entirely sure how much of that ended up in the package documentation).

In the perfect world, a package URI would be "package://path/something" with the first "path" not being part of the hierarchical path. That's how we try to treat it anyway.

That means that "package:file" is not specifying a directory. The "file" is not a directory, it's a package selector. We could default that to meaning the same as "package:file/", but it is a different URI (read as plain URIs, the former is a file, the latter is a directory).

**Comment by nex3** [comment]
> In the perfect world, a package URI would be "package://path/something" with the first "path" not being part of the hierarchical path. That's how we try to treat it anyway.
>
> That means that "package:file" is not specifying a directory. The "file" is not a directory, it's a package selector. We could default that to meaning the same as "package:file/", but it is a different URI (read as plain URIs, the former is a file, the latter is a directory).

If we're using an analogy to more common URI schemes, I'd expect `package:path` to work the same as `http://host`, which is resolved the same as `http://host`.

> If you do `Uri.parse("package:file").resolve("file.dart")`, you get the URI `package:file.dart`. 
If you do `Uri.parse("package:file").resolve("/file.dart")`, you get the URI `package:/file.dart`. 
It just doesn't work very well without the slash.  

But you said above that you try to treat `package:path` as `package://path`. `Uri.parse("package://path").resolve("file.dart")` returns `package://path/file.dart`.

All this debate over URI semantics is kind of beside the point, though. It's clear what the user means when they write `package:path`. It costs nothing to make it work. That should be enough to do so.

**Comment by kasperl** [comment]
Crash in VM builtin --> @iposva-google.

**Comment by iposva-google** [comment]
Out for review: https://codereview.chromium.org/1822863002

**Comment by iposva-google** [comment]
Fixed with 3b2a695d71057f3a538f79f48a96f6bdca2474bb
